### PR TITLE
feat: System Implementation for RL adapters (Gymnasium, PettingZoo, PufferLib)

### PR DESCRIPTION
**Pull request by teodorionita**
# Complete Environment Adapter System and PettingZoo Integration

## Summary

Implemented comprehensive environment adapter system with **four framework adapters** (PettingZoo, Gym, PufferLib, Core C++) and established clean architectural separation between primary training and external research compatibility layers.

### Complete Adapter System Implementation

- **PettingZoo Adapter**: `mettagrid/src/metta/mettagrid/pettingzoo_wrapper.py`
  - Full Parallel API compliance with agent dictionaries
  - Verified with `pettingzoo.test.api_test`
  - Proper agent naming (`agent_0`, `agent_1`, etc.)
  
- **Gymnasium Adapter**: `mettagrid/src/metta/mettagrid/gym_wrapper.py`
  - Pure Gymnasium interface for external research
  - Multi-agent Tuple spaces or single-agent Box spaces
  
- **PufferLib Adapter**: `mettagrid/src/metta/mettagrid/puffer_wrapper.py`
  - Lightweight PufferLib compatibility layer
  - Enables external PufferLib training systems
  
- **Core C++ Interface**: `mettagrid/src/metta/mettagrid/core.py`
  - Direct C++ MettaGrid wrapper with explicit buffer management
  - Stateless interface for maximum performance
  
- **Base Environment**: `mettagrid/src/metta/mettagrid/base_env.py`
  - Abstract base class providing common functionality
  - Curriculum, stats, and replay management for all adapters

### Architecture Clarification  
- **Primary Environment**: `MettaGridEnv` - actively developed for Softmax Studio training
- **External Adapters**: Lightweight compatibility layers for external research frameworks
  - `MettaGridGymEnv` - Pure Gymnasium interface
  - `MettaGridPettingZooEnv` - PettingZoo multi-agent interface  
  - `MettaGridPufferEnv` - PufferLib compatibility layer
- **Trainer Validation**: Preserved existing validation ensuring main training uses `MettaGridEnv` only

### Testing & Integration

- **Comprehensive Test Suite**: 
  - `mettagrid/tests/test_pettingzoo_wrapper.py` with PettingZoo API compliance
  - `mettagrid/tests/test_gym_wrapper.py` for Gymnasium compatibility
  - `mettagrid/tests/test_puffer_wrapper.py` for PufferLib integration
  - `mettagrid/tests/test_core.py` for C++ interface
- **Demo Integration**: Complete demo suite for all four adapters
  - `demo_train_pettingzoo.py` - PettingZoo multi-agent research
  - `demo_train_gym.py` - Gymnasium compatibility 
  - `demo_train_puffer.py` - External PufferLib training
  - `demo_train_core.py` - Direct C++ interface
- **CI Automation**: Automated testing for PettingZoo, Puffer, and Core demos in GitHub Actions
- **Training Validation**: Main training system (`tools/train.py`) works with `MettaGridEnv`
- **CI Pipeline**: All 949 tests pass, complete framework compatibility verified

### Documentation
- **README Updates**: Added Environment Architecture section explaining design philosophy
- **Usage Examples**: Documented adapter usage patterns for external researchers
- **API Reference**: Complete PettingZoo wrapper API documentation

## Architecture
```
Main Training:     tools/train.py → MettaTrainer → MettaGridEnv → MettaGridCore (C++)

External Research: External Code → [Framework Adapter] → MettaGridCore (C++)
                                  ↓
                   MettaGridGymEnv (Gymnasium)
                   MettaGridPettingZooEnv (PettingZoo)  
                   MettaGridPufferEnv (PufferLib)
```

## Results
- **PettingZoo Compatibility**: Full Parallel API support enabling MARL research integration
- **Zero Breaking Changes**: All existing training code unchanged
- **Clear Architecture**: Separated main training from external framework compatibility
- **Performance Preserved**: C++ core maintains high performance across all adapters
- **Research Enablement**: External researchers can use MettaGrid with their preferred frameworks

## What Was NOT Changed
- **Core Training**: `MettaTrainer` and `MettaGridEnv` remain primary training system
- **C++ Performance**: Original MettaGrid C++ implementation preserved
- **Existing APIs**: All base environment methods and training interfaces unchanged
- **Validation Logic**: Trainer continues to enforce `MettaGridEnv` for main training

## Dependencies

- **Existing**: `pettingzoo>=1.25.0` (already in pyproject.toml)
- **No New Dependencies**: Leveraged existing MettaGrid infrastructure

### Trainer-Agnostic Core (Not Implemented)

**Current State**: `MettaGridEnv` inherits from `PufferEnv, GymEnv` making it trainer-specific
**Future Goal**: Make core trainer-agnostic with adapter pattern

```text
Proposed: MettaGridCore → [Trainer Adapters] → Different Training Systems
          Currently: MettaGridEnv (PufferEnv+GymEnv) → MettaTrainer only
```

**Requirements for Future Implementation**:

- Switch alias or flag: Make `MettaGridEnv` point to MettaTrainer adapter, not PufferLib inheritance
- Expose `env_api` parameter to be explicit about trainer interface
- Enhanced CI coverage: Currently test demos per adapter, future needs trainer × adapter matrix

### Open Questions for Team Review

1. **Adapter Architecture**: Should we refactor `MettaGridEnv` to be a configurable wrapper rather than a fixed inheritance?
2. **Training System Independence**: Should we implement the full trainer-agnostic pattern to allow future divergence?
3. **Enhanced CI Matrix**: Should we expand from current per-adapter testing to full trainer × adapter matrix testing?
4. **C++ Buffer Cleanup**: Remove unused internal buffers from C++ core for cleaner architecture (planned for future)

---

**Status**: Complete environment adapter ecosystem delivered - four framework adapters (PettingZoo, Gym, PufferLib, Core C++) with comprehensive testing and CI coverage. External adapters serve as research compatibility layers while main training system remains focused on `MettaGridEnv`. Future work identified for trainer-agnostic architecture.

[Asana Task](https://app.asana.com/1/1209016784099267/project/1210348820405981/task/1210859343854928)

[Asana Task](https://app.asana.com/1/1209016784099267/project/1210348820405981/task/1210863452234612)

### CODE REVIEW

File: metta/api.py [hunk 1]
```diff
@@ -746,7 +745,11 @@ def __new__(
 
         # Get the actual MettaGridEnv from vecenv wrapper
         metta_grid_env = env.driver_env
-        assert isinstance(metta_grid_env, MettaGridEnv)
+        # Check if driver_env is one of our valid MettaGrid adapters
+        from metta.mettagrid import MettaGridCore, MettaGridGymEnv, MettaGridPettingZooEnv, MettaGridPufferEnv
```
Comment (rjwalters) [inline metta/api.py:749]: I think for our metta trainer only MettaGridEnv should be valid?

File: mettagrid/README.md [hunk 2]
```diff
@@ -139,6 +139,51 @@ layout of the gridworld, the placement of objects, and various properties of the
 - Cooldown between uses (varies by building)
 - Can be damaged and destroyed by attacks
 
+## Environment Architecture
+
+MettaGrid features a modular environment architecture that supports multiple RL frameworks:
+
+### Environment Adapters
+
+- **`MettaGridEnv`** - Main environment class (backward compatible with PufferLib)
```
Comment (rjwalters) [inline mettagrid/README.md:148]:
```suggestion
- **`MettaGridEnv`** - Main environment class for use with Softmax Studio
```

File: mettagrid/README.md [hunk 3]
```diff
@@ -139,6 +139,51 @@ layout of the gridworld, the placement of objects, and various properties of the
...
+# Test PettingZoo adapter with training
+python -m mettagrid.demos.demo_train_pettingzoo
+
+# Test Puffer adapter with training  
+python -m mettagrid.demos.demo_train_puffer
+
+# Test Gym adapter with training
+python -m mettagrid.demos.demo_train_gym
+
+# Test Core adapter with training
```
Comment (rjwalters) [inline mettagrid/README.md:176]: I don't expect that we will have a trainer that directly uses Core -- we should probably separate out a list of "external training system adapters" and not include code on it

File: mettagrid/demos/demo_train_puffer.py [hunk 4]
```diff
@@ -0,0 +1,341 @@
+#!/usr/bin/env python3
+# /// script
+# requires-python = ">=3.10"
+# dependencies = [
```
Comment (rjwalters) [inline mettagrid/demos/demo_train_puffer.py:4]: I would expect this to depend on pufferlib?

File: mettagrid/src/metta/mettagrid/mettagrid_env.py [hunk 5]
```diff
@@ -1,506 +1,90 @@
...
-    @property
-    def inventory_item_names(self) -> list[str]:
-        return self._c_env.inventory_item_names()
-
-    @property
-    def initial_grid_hash(self) -> int:
-        """Returns the hash of the initial grid configuration."""
-        return self._c_env.initial_grid_hash
+# Adapter mapping for configurable environment selection (training-compatible adapters only)
+_ADAPTER_MAP = {
```
Comment (rjwalters) [inline mettagrid/src/metta/mettagrid/mettagrid_env.py:39]: I think we should instead force people to init the environment they want -- this will keep the softmax version cleaner

File: mettagrid/README.md (hunk 3 above)
Comment (teodorionita) [reply mettagrid/README.md]: I've kept the demo for core still running just to be safe but removed it from readme completely

File: metta/api.py (hunk 1 above)
Comment (teodorionita) [reply metta/api.py]: fixed

File: mettagrid/demos/demo_train_gym.py [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+        duration = time.time() - start_time
+        print("\n" + "=" * 60)
+        print("GYM TRAINING INTEGRATION COMPLETED")
+        print("=" * 60)
+        print("Gym adapter functionality: Works correctly")
+        print("Multi-agent training: Compatible with training pipeline")
+        print("Training integration: Short training run successful")
+        print(f"\nTotal test time: {duration:.1f} seconds")
+        print("\nGym adapter is ready for production training")
+        print("=" * 60)
```
[the remark this answers is not in the document]
Comment (teodorionita) [reply mettagrid/demos/demo_train_gym.py:304]: demo, no need for this

File: .github/workflows/test-demo-environments.yml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+          echo "TIMEOUT: ${{ matrix.demo }} ran for 45 seconds (expected completion ~5s)"
+          exit 1
+        elif [ $exit_code -eq 0 ]; then
+          echo "SUCCESS: ${{ matrix.demo }} completed successfully"
+        else
+          echo "FAILED: ${{ matrix.demo }} failed with exit code $exit_code"
+          exit 1
+        fi
+    
+    - name: Run linting on demo files
```
Comment (rjwalters) [inline .github/workflows/test-demo-environments.yml:49]: this is not needed - we check linting on the full repo in Checks

File: metta/sim/simulation.py [hunk 8]
```diff
@@ -159,7 +159,10 @@ def __init__(
         self._stats_epoch_id: uuid.UUID | None = stats_epoch_id
 
         metta_grid_env: MettaGridEnv = self._vecenv.driver_env  # type: ignore
-        assert isinstance(metta_grid_env, MettaGridEnv)
```
Comment (rjwalters) [inline metta/sim/simulation.py:162]: same - no need to change simulation.py here

Comment (teodorionita) [reply metta/sim/simulation.py]: reverted to main version

File: mettagrid/src/metta/mettagrid/core.py [hunk 9]
```diff
@@ -55,6 +55,9 @@ def __init__(
         self._map_grid = map_grid
         self._seed = seed
 
+        # Add PufferLib compatibility attributes
+        self.emulated = False  # Required by PufferLib vectorization
```
Comment (rjwalters) [inline mettagrid/src/metta/mettagrid/core.py:59]: should this be in core or just the puffer env?

Comment (teodorionita) [reply mettagrid/src/metta/mettagrid/core.py:59]: good catch should only be in puffer_env.py, moving it now
